oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,284 @@
+            syncConfigFor(simName)
+        }
+    }
+
+    override fun cancelAll() {
+        GlobalScope.launch {
+            for ((name, sim) in modelService.simByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    override fun cancel(name: String) {
+        GlobalScope.launch {
+            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+            ).build()
+
+            modelService.simByName.getValue(name).input.onNext(message)
+        }
+    }
+
+    override fun disconnectAll() {
+       throw NotImplementedException()
+    }
+
+    override fun cancelAndStop() {
+        GlobalScope.launch {
+            stopOptimization()
+            for ((name, sim) in modelService.simByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    private suspend fun syncConfigFor(simName: String) {
+        val sim = modelService.simByName.getValue(simName)
+        val message = OASISQueryDTO.newBuilder().setNodeStatusRequest(
+                OASISQueryDTO.NodeStatusUpdateRequest.newBuilder().setName(simName)
+        ).build()
+
+        sim.input.onNext(message)
+
+        val result = select<Either<Simulation, Message>> {
+            sim.update.onReceive { Either.Left(updateFromResponse(it)) }
+            sim.error.onReceive {
+                Either.Right(Message(it.name, "Error update simulation $simName due to ${it.message} :\n${it.exception}"))
+            }
+            onTimeout(Duration.ofSeconds(5).toMillis()) {
+                Either.Right(Message(simName, "Update simulation timeout. Please check simulation is registered and reponsive."))
+            }
+        }
+
+        if (result.isLeft()) {
+            modelService.updateSim(simName, result.left().get())
+        } else {
+            modelService.addMessage(result.right().get())
+        }
+
+    }
+
+    override fun stopOptimization(request: StopOptimizationCommandDTO, responseObserver: StreamObserver<StopOptimizationResponseDTO>) = responseObserver.consume {
+        stopOptimization()
+        StopOptimizationResponseDTO.newBuilder().setMessage("Stopped").setStopped(true).build()
+    }
+
+    override fun offerSimulationResult(request: SimulationResponseDTO, responseObserver: StreamObserver<SimulationResultConfirmDTO>) = responseObserver.consume {
+        val output = modelService.simByName[request.name]?.output
+        if (output != null) {
+            output.offer(request)
+        } else {
```

Comment:
yeah so, I'm pretty sure I said this elsewhere but Ill say it again here.

**this strategy is fundamentally not safe**

we cant call `offer` like this. It means that if our buffer fills up (and by default we have buffers of size 1), then we start dropping requests.

on the `error` channel this will be particularly bad.